### PR TITLE
WT-7912 Fix prefix search near optimisation where key range is split across pages

### PR DESCRIPTION
**Pull request by mm-ng**

### CODE REVIEW

File: src/btree/bt_curnext.c [hunk 1]
```diff
@@ -334,8 +334,8 @@ __cursor_var_next(WT_CURSOR_BTREE *cbt, bool newpage, bool restart, size_t *skip
  *     Move to the next row-store item.
  */
 static inline int
-__cursor_row_next(
-  WT_CURSOR_BTREE *cbt, bool newpage, bool restart, size_t *skippedp, WT_ITEM *prefix)
+__cursor_row_next(WT_CURSOR_BTREE *cbt, bool newpage, bool restart, bool *prefix_key_out_of_bounds,
```
Comment (etienneptl) [inline src/btree/bt_curnext.c:337]: Should we have `bool *prefix_key_out_of_bounds` as the last arg after `WT_ITEM *prefix`?

File: src/btree/bt_curnext.c [hunk 2]
```diff
@@ -750,13 +754,14 @@ __wt_btcur_next_prefix(WT_CURSOR_BTREE *cbt, WT_ITEM *prefix, bool truncating)
...
                 total_skipped += skipped;
                 /*
-                 * We can directly return WT_NOTFOUND here as the caller expects the cursor to be
-                 * positioned when traversing keys for prefix search near.
+                 * If we are doing a search near with prefix key configured, we need to check if we
+                 * have exited the cursor row next function due to a prefix key mismatch. If so, we
+                 * can immediately return WT_NOTFOUND and we do not have to walk onto the next page.
                  */
-                if (ret == WT_NOTFOUND && F_ISSET(&cbt->iface, WT_CURSTD_PREFIX_SEARCH))
+                if (ret == WT_NOTFOUND && prefix_key_out_of_bounds)
```
Comment (etienneptl) [inline src/btree/bt_curnext.c:764]: Do we still need to check for `ret == WT_NOTFOUND`? We could assert if we wanted to keep it but given where `prefix_key_out_of_bounds` is set to true, ret should be `WT_NOTFOUND`.

File: src/btree/bt_curnext.c [hunk 3]
```diff
@@ -750,13 +753,14 @@ __wt_btcur_next_prefix(WT_CURSOR_BTREE *cbt, WT_ITEM *prefix, bool truncating)
                 total_skipped += skipped;
                 break;
             case WT_PAGE_ROW_LEAF:
-                ret = __cursor_row_next(cbt, newpage, restart, &skipped, prefix);
+                ret = __cursor_row_next(cbt, newpage, &prefix_key_oob, restart, &skipped, prefix);
```
[the remark this answers is not in the document]
Comment (etienneptl) [reply src/btree/bt_curnext.c]: I agree, should be at the end, I raised a similar comment.

File: src/btree/bt_curnext.c (hunk 2 above)
Comment (luke-pearson) [reply src/btree/bt_curnext.c:764]: I agree with an assert here.

File: src/btree/bt_curnext.c (hunk 1 above)
Comment (mm-ng) [reply src/btree/bt_curnext.c:337]: I thought the prefix_key_out_of_bounds variable was already at the end with all the other pointers, but I can put it as the very last variable if that's the standard?

File: src/btree/bt_curnext.c (hunk 2 above)
Comment (mm-ng) [reply src/btree/bt_curnext.c:764]: Ok, I will probably use a WT_ASSERT here instead.

File: src/btree/bt_curnext.c (hunk 1 above)
Comment (etienneptl) [reply src/btree/bt_curnext.c:337]: I usually put the outputs at the end of the function args, I think `prefix` which is currently at the end is an input.
@quchenhao, is the current place of `prefix_key_out_of_bounds` correct?

File: src/btree/bt_cursor.c [hunk 4]
```diff
@@ -740,10 +740,22 @@ __wt_btcur_search_near(WT_CURSOR_BTREE *cbt, int *exactp)
...
+         *
+         * For example, if we search with the prefix "b", the cursor will be positioned at the first
+         * key starting with "b". All the entries before this one will start with "a", hence not
+         * matching the prefix.
+         */
+        if (F_ISSET(cursor, WT_CURSTD_PREFIX_SEARCH))
+            goto done;
+
         /*
          * We walked to the end of the tree without finding a match. Walk backwards instead.
```
Comment (etienneptl) [inline src/btree/bt_cursor.c:756]: @luke-pearson, question that is out of the scope of the PR but probably worth it:
Why do we walk backwards from the last position of the cursor? Why don't we save the initial position of the cursor and start from there and go backwards?

Comment (luke-pearson) [reply src/btree/bt_cursor.c:756]: That's a good question and I remember i brought it up with Keith B but I forget the actual answer, apparently its not that trivial to do that?

Comment (etienneptl) [reply src/btree/bt_cursor.c:756]: I imagined it was not trivial, I will ask him and let you know!

Comment (mm-ng) [reply src/btree/bt_cursor.c:756]: Oooh, I'd be interested to know as well if you do get an answer on that

Comment (etienneptl) [reply src/btree/bt_cursor.c:756]: The answer is that the key could be gone by the time we restart the search, that's why we cannot do it.

File: src/btree/bt_curnext.c (hunk 1 above)
Comment (mm-ng) [reply src/btree/bt_curnext.c:337]: I will move it to the last argument, that is what I see in other code samples

**Review by etienneptl — APPROVED** [review]
LGTM after the last changes!

**Review by luke-pearson — APPROVED** [review]
LGTM.

**Review by quchenhao — APPROVED** [review]
lgtm

**Pull request merged**